### ISSUE TITLE
frag() error in iri.t on NetBSD/amd64

### ISSUE THREAD

**Issue by sysread** [sysread/URI-Fast#13]
From [cpan testers report](https://www.cpantesters.org/cpan/report/eb90f324-7b6a-11e8-a944-5f7713bf8fb6):

```"/home/cpan/pit/bare/perl-5.18.0/bin/perl" -MExtUtils::Command::MM -e 'cp_nonempty' -- Fast.bs blib/arch/auto/URI/Fast/Fast.bs 644
PERL_DL_NONLAZY=1 "/home/cpan/pit/bare/perl-5.18.0/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*
t/basics.t ........ ok
t/compare.t ....... ok
t/encoding.t ...... ok
t/ipv.t ........... ok
    # Failed test 'frag'
    # at t/iri.t line 22.
    # Unicode::GCString is not installed, table may not display all unicode characters properly
    # +----------------+----+-------+
    # | GOT            | OP | CHECK |
    # +----------------+----+-------+
    # | ÃÂ\N{U+92}ÃÂÃÂ®ÃÂÃÂ¥ÃÂ | eq | ÃÂÃÂ®ÃÂ¥ÃÂ©  |
    # +----------------+----+-------+

# Failed test 'getters'
# at t/iri.t line 27.
# Seeded srand with seed '20180629' from local date.
t/iri.t ........... 
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/4 subtests 
t/memory.t ........ ok
t/param.t ......... ok
t/parsing.t ....... ok
t/query_keyset.t .. ok
t/test.t .......... ok
t/tied.t .......... ok

Test Summary Report
-------------------
t/iri.t         (Wstat: 256 Tests: 4 Failed: 1)
  Failed test:  3
  Non-zero exit status: 1
Files=11, Tests=60,  2 wallclock secs ( 0.09 usr  0.03 sys +  1.68 cusr  0.55 csys =  2.35 CPU)
Result: FAIL
Failed 1/11 test programs. 1/60 subtests failed.
*** Error code 255

Stop.
make: stopped in /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aEM_moZtOc/URI-Fast-0.35


PREREQUISITES:

Here is a list of prerequisites you specified and versions we
managed to load:

	  Module Name                        Have     Want
	  Carp                               1.29        0
	  Exporter                           5.68        0
	  ExtUtils::MakeMaker                7.34  6.63_03
	  ExtUtils::testlib                  7.34        0
	  Test2                          1.302136 1.302125
	  Test2::Suite                   0.000114 0.000100
	  Test2::V0                      0.000114        0
	  Test::LeakTrace                    0.16     0.16
	  Test::Pod                          1.52     1.41
	  URI::Encode::XS                    0.11     0.11
	  URI::Split                         1.74        0
	  parent                            0.225        0

Perl module toolchain versions installed:
	Module Name                        Have
	CPANPLUS                         0.9174
	CPANPLUS::Dist::Build              0.88
	Cwd                                3.74
	ExtUtils::CBuilder             0.280230
	ExtUtils::Command                  7.34
	ExtUtils::Install                  2.14
	ExtUtils::MakeMaker                7.34
	ExtUtils::Manifest                 1.71
	ExtUtils::ParseXS                  3.35
	File::Spec                         3.74
	Module::Build                    0.4224
	Pod::Parser                        1.60
	Pod::Simple                        3.28
	Test2                          1.302136
	Test::Harness                      3.42
	Test::More                     1.302136
	version                          0.9924

******************************** NOTE ********************************
The comments above are created mechanically, possibly without manual
checking by the sender.  As there are many people performing automatic
tests on each upload to CPAN, it is likely that you will receive
identical messages about the same problem.

If you believe that the message is mistaken, please reply to the first
one with correction and/or additional informations, and do not take
it personally.  We appreciate your patience. :)
**********************************************************************

Additional comments:


This report was machine-generated by CPANPLUS::Dist::YACSmoke 1.02.
Powered by minismokebox version 0.68

------------------------------
ENVIRONMENT AND OTHER CONTEXT
------------------------------

Environment variables:

    AUTOMATED_TESTING = 1
    NONINTERACTIVE_TESTING = 1
    PATH = /home/cpan/bin:/bin:/sbin:/usr/bin:/usr/sbin:/usr/X11R7/bin:/usr/X11R6/bin:/usr/pkg/bin:/usr/pkg/sbin:/usr/games:/usr/local/bin:/usr/local/sbin
    PERL5LIB = :/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aUVR3uqHM3/Importer-0.025/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aUVR3uqHM3/Importer-0.025/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/z8T7vtmLr3/Scope-Guard-0.21/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/z8T7vtmLr3/Scope-Guard-0.21/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/IvdCuuczxa/Sub-Info-0.002/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/IvdCuuczxa/Sub-Info-0.002/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aeMieONlaW/Term-Table-0.012/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aeMieONlaW/Term-Table-0.012/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/oKcMbBBr_o/Test2-Suite-0.000114/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/oKcMbBBr_o/Test2-Suite-0.000114/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/JsVUQLwnlV/Test-LeakTrace-0.16/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/JsVUQLwnlV/Test-LeakTrace-0.16/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/UPexbq5kGM/Test-Pod-1.52/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/UPexbq5kGM/Test-Pod-1.52/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/YtPLHS0AYC/URI-Encode-XS-0.11/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/YtPLHS0AYC/URI-Encode-XS-0.11/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/X_WHyUoOeo/Test-Needs-0.002005/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/X_WHyUoOeo/Test-Needs-0.002005/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/h6nCoXWXd1/URI-1.74/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/h6nCoXWXd1/URI-1.74/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aEM_moZtOc/URI-Fast-0.35/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aEM_moZtOc/URI-Fast-0.35/blib/arch
    PERL5_CPANPLUS_IS_RUNNING = 24578
    PERL5_CPANPLUS_IS_VERSION = 0.9174
    PERL5_MINISMOKEBOX = 0.68
    PERL5_YACSMOKE_BASE = /home/cpan/pit/bare/conf/perl-5.18.0
    PERL_EXTUTILS_AUTOINSTALL = --defaultdeps
    PERL_MM_USE_DEFAULT = 1
    SHELL = /usr/pkg/bin/bash
    TERM = screen

Perl special variables (and OS-specific diagnostics, for MSWin32):

    Perl: $^X = /home/cpan/pit/bare/perl-5.18.0/bin/perl
    UID:  $<  = 1002
    EUID: $>  = 1002
    GID:  $(  = 100 100
    EGID: $)  = 100 100


-------------------------------


--

Summary of my perl5 (revision 5 version 18 subversion 0) configuration:
   
  Platform:
    osname=netbsd, osvers=6.1.4, archname=amd64-netbsd
    uname='netbsd naboo.bingosnet.co.uk 6.1.4 netbsd 6.1.4 (generic) amd64 '
    config_args='-des -Dprefix=/home/cpan/pit/bare/perl-5.18.0'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/pkg/include',
    optimize='-O',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/pkg/include'
    ccversion='', gccversion='4.5.3', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -Wl,-rpath,/usr/pkg/lib -Wl,-rpath,/usr/local/lib -fstack-protector -L/usr/pkg/lib'
    libpth=/usr/pkg/lib /lib /usr/lib
    libs=-lgdbm -lm -lcrypt -lutil -lc -lposix
    perllibs=-lm -lcrypt -lutil -lc -lposix
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E '
    cccdlflags='-DPIC -fPIC ', lddlflags='--whole-archive -shared  -L/usr/pkg/lib -fstack-protector'


Characteristics of this binary (from libperl): 
  Compile-time options: HAS_TIMES PERLIO_LAYERS PERL_DONT_CREATE_GVSV
                        PERL_HASH_FUNC_ONE_AT_A_TIME_HARD PERL_MALLOC_WRAP
                        PERL_PRESERVE_IVUV PERL_SAWAMPERSAND USE_64_BIT_ALL
                        USE_64_BIT_INT USE_LARGE_FILES USE_LOCALE
                        USE_LOCALE_COLLATE USE_LOCALE_CTYPE
                        USE_LOCALE_NUMERIC USE_PERLIO USE_PERL_ATOF
  Built under netbsd
  Compiled at May  9 2014 16:48:11
  %ENV:
    PERL5LIB=":/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aUVR3uqHM3/Importer-0.025/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aUVR3uqHM3/Importer-0.025/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/z8T7vtmLr3/Scope-Guard-0.21/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/z8T7vtmLr3/Scope-Guard-0.21/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/IvdCuuczxa/Sub-Info-0.002/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/IvdCuuczxa/Sub-Info-0.002/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aeMieONlaW/Term-Table-0.012/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aeMieONlaW/Term-Table-0.012/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/oKcMbBBr_o/Test2-Suite-0.000114/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/oKcMbBBr_o/Test2-Suite-0.000114/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/JsVUQLwnlV/Test-LeakTrace-0.16/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/JsVUQLwnlV/Test-LeakTrace-0.16/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/UPexbq5kGM/Test-Pod-1.52/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/UPexbq5kGM/Test-Pod-1.52/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/YtPLHS0AYC/URI-Encode-XS-0.11/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/YtPLHS0AYC/URI-Encode-XS-0.11/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/X_WHyUoOeo/Test-Needs-0.002005/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/X_WHyUoOeo/Test-Needs-0.002005/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/h6nCoXWXd1/URI-1.74/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/h6nCoXWXd1/URI-1.74/blib/arch:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aEM_moZtOc/URI-Fast-0.35/blib/lib:/home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aEM_moZtOc/URI-Fast-0.35/blib/arch"
    PERL5_CPANPLUS_IS_RUNNING="24578"
    PERL5_CPANPLUS_IS_VERSION="0.9174"
    PERL5_MINISMOKEBOX="0.68"
    PERL5_YACSMOKE_BASE="/home/cpan/pit/bare/conf/perl-5.18.0"
    PERL_EXTUTILS_AUTOINSTALL="--defaultdeps"
    PERL_MM_USE_DEFAULT="1"
  @INC:
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aUVR3uqHM3/Importer-0.025/blib/lib
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aUVR3uqHM3/Importer-0.025/blib/arch
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/z8T7vtmLr3/Scope-Guard-0.21/blib/lib
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/z8T7vtmLr3/Scope-Guard-0.21/blib/arch
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/IvdCuuczxa/Sub-Info-0.002/blib/lib
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/IvdCuuczxa/Sub-Info-0.002/blib/arch
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aeMieONlaW/Term-Table-0.012/blib/lib
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aeMieONlaW/Term-Table-0.012/blib/arch
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/oKcMbBBr_o/Test2-Suite-0.000114/blib/lib
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/oKcMbBBr_o/Test2-Suite-0.000114/blib/arch
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/JsVUQLwnlV/Test-LeakTrace-0.16/blib/lib
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/JsVUQLwnlV/Test-LeakTrace-0.16/blib/arch
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/UPexbq5kGM/Test-Pod-1.52/blib/lib
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/UPexbq5kGM/Test-Pod-1.52/blib/arch
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/YtPLHS0AYC/URI-Encode-XS-0.11/blib/lib
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/YtPLHS0AYC/URI-Encode-XS-0.11/blib/arch
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/X_WHyUoOeo/Test-Needs-0.002005/blib/lib
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/X_WHyUoOeo/Test-Needs-0.002005/blib/arch
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/h6nCoXWXd1/URI-1.74/blib/lib
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/h6nCoXWXd1/URI-1.74/blib/arch
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aEM_moZtOc/URI-Fast-0.35/blib/lib
    /home/cpan/pit/bare/conf/perl-5.18.0/.cpanplus/5.18.0/build/aEM_moZtOc/URI-Fast-0.35/blib/arch
    /home/cpan/pit/bare/perl-5.18.0/lib/site_perl/5.18.0/amd64-netbsd
    /home/cpan/pit/bare/perl-5.18.0/lib/site_perl/5.18.0
    /home/cpan/pit/bare/perl-5.18.0/lib/5.18.0/amd64-netbsd
    /home/cpan/pit/bare/perl-5.18.0/lib/5.18.0
    .```

**Comment by sysread** [comment]
I believe the root cause of this may be the size of the frag (and other) struct members. The sizes may be reasonable for ASCII, but not necessarily for multi-bytes sequences in IRIs. In `uri_scan` and `uri_scan_auth`, segments are copied into the struct using `strncpy` with the read length set to the minimum of the number of bytes found using `strcspn` and the length of the member. In cases where this would overflow with an ASCII value, the value is simply cut off. In cases where this overflows with a multi-byte utf8 sequence, it may produce an invalid sequence that cannot be decoded.

**Comment by sysread** [comment]
This is also happening on win32.

**Comment by sysread** [comment]
The root issue appears to be that the input SV* is not consistently flagged as utf8 by various platform perls. I've pushed 0.35_4 to cpan to verify this by encoding unflagged inputs before calling `uri_scan`.

http://matrix.cpantesters.org/?dist=URI-Fast+0.35_4

**Comment by sysread** [comment]
This is fixed in 0.35_7, at least, and will be in 0.36.
